File: animated-drawable/src/test/java/com/facebook/fresco/animation/bitmap/preparation/FixedNumberBitmapFramePreparationStrategyTest.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.fresco.animation.bitmap.preparation;

import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.inOrder;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import com.facebook.fresco.animation.backend.AnimationBackend;
import com.facebook.fresco.animation.bitmap.BitmapFrameCache;
import kotlin.Unit;
import kotlin.jvm.functions.Function0;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InOrder;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;

/** Tests {@link FixedNumberBitmapFramePreparationStrategy}. */
@RunWith(RobolectricTestRunner.class)
public class FixedNumberBitmapFramePreparationStrategyTest {

  private static final int NUMBER_OF_FRAMES_TO_PREPARE = 3;
  private static final int FRAME_COUNT = 10;

  @Mock public AnimationBackend mAnimationBackend;
  @Mock public BitmapFramePreparer mBitmapFramePreparer;
  @Mock public BitmapFrameCache mBitmapFrameCache;
  @Mock public Function0<Unit> onAnimationLoaded;

  private BitmapFramePreparationStrategy mBitmapFramePreparationStrategy;

  @Before
  public void setup() {
    MockitoAnnotations.initMocks(this);
    mBitmapFramePreparationStrategy =
        new FixedNumberBitmapFramePreparationStrategy(NUMBER_OF_FRAMES_TO_PREPARE);
    when(mAnimationBackend.getFrameCount()).thenReturn(FRAME_COUNT);
    when(mBitmapFramePreparer.prepareFrame(eq(mBitmapFrameCache), eq(mAnimationBackend), anyInt()))
        .thenReturn(true);
  }

  @Test
  public void testPrepareFrames_FromFirstFrame() throws Exception {
    mBitmapFramePreparationStrategy.prepareFrames(
        mBitmapFramePreparer, mBitmapFrameCache, mAnimationBackend, 0, onAnimationLoaded);
    verifyPrepareCalledForFramesInOrder(1, 2, 3);
  }

  @Test
  public void testPrepareFrames_FromLastFrame() throws Exception {
    mBitmapFramePreparationStrategy.prepareFrames(
        mBitmapFramePreparer, mBitmapFrameCache, mAnimationBackend, 9, onAnimationLoaded);
    verifyPrepareCalledForFramesInOrder(0, 1, 2);
  }

  @Test
  public void testPrepareFrames_ExactlyLastFrames() throws Exception {
    mBitmapFramePreparationStrategy.prepareFrames(
        mBitmapFramePreparer, mBitmapFrameCache, mAnimationBackend, 6, onAnimationLoaded);
    verifyPrepareCalledForFramesInOrder(7, 8, 9);
  }

  @Test
  public void testPrepareFrames_FrameOverflow() throws Exception {
    mBitmapFramePreparationStrategy.prepareFrames(
        mBitmapFramePreparer, mBitmapFrameCache, mAnimationBackend, 8, onAnimationLoaded);
    verifyPrepareCalledForFramesInOrder(9, 0, 1);
  }

  @Test
  public void testPrepareFrames_FromFirstFrame_WhenBitmapFramePreparerAlwaysFails()
      throws Exception {
    when(mBitmapFramePreparer.prepareFrame(eq(mBitmapFrameCache), eq(mAnimationBackend), anyInt()))
        .thenReturn(false);
    mBitmapFramePreparationStrategy.prepareFrames(
        mBitmapFramePreparer, mBitmapFrameCache, mAnimationBackend, 0, onAnimationLoaded);
    verifyPrepareCalledForFramesInOrder(1);
  }

  @Test
  public void testPrepareFrames_FromFirstFrame_WhenBitmapFramePreparerFailsForSelectedFrames()
      throws Exception {
    when(mBitmapFramePreparer.prepareFrame(eq(mBitmapFrameCache), eq(mAnimationBackend), eq(2)))
        .thenReturn(false);
    when(mBitmapFramePreparer.prepareFrame(eq(mBitmapFrameCache), eq(mAnimationBackend), eq(3)))
        .thenReturn(false);
    mBitmapFramePreparationStrategy.prepareFrames(
        mBitmapFramePreparer, mBitmapFrameCache, mAnimationBackend, 0, onAnimationLoaded);
    verifyPrepareCalledForFramesInOrder(1, 2);
  }

  @Test
  public void testPrepareFrames_onAnimationLoadedIsTrigger_WhenFramesAreLoaded() {
    mBitmapFramePreparationStrategy.prepareFrames(
        mBitmapFramePreparer, mBitmapFrameCache, mAnimationBackend, 0, onAnimationLoaded);

    verify(onAnimationLoaded).invoke();
  }

  private void verifyPrepareCalledForFramesInOrder(int... frameNumbers) {
    InOrder inOrderBitmapFramePreparer = inOrder(mBitmapFramePreparer);
    for (int frameNumber : frameNumbers) {
      inOrderBitmapFramePreparer
          .verify(mBitmapFramePreparer)
          .prepareFrame(mBitmapFrameCache, mAnimationBackend, frameNumber);
    }
    inOrderBitmapFramePreparer.verifyNoMoreInteractions();
  }
}

File: animated-drawable/src/test/java/com/facebook/fresco/animation/bitmap/preparation/FixedNumberBitmapFramePreparationStrategyTest.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.fresco.animation.bitmap.preparation

import com.facebook.fresco.animation.backend.AnimationBackend
import com.facebook.fresco.animation.bitmap.BitmapFrameCache
import kotlin.Unit
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.MockitoAnnotations
import org.mockito.kotlin.any
import org.mockito.kotlin.eq
import org.mockito.kotlin.inOrder
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever
import org.robolectric.RobolectricTestRunner

/** Tests [FixedNumberBitmapFramePreparationStrategy]. */
@RunWith(RobolectricTestRunner::class)
class FixedNumberBitmapFramePreparationStrategyTest {

  companion object {
    const val NUMBER_OF_FRAMES_TO_PREPARE = 3
    const val FRAME_COUNT = 10
  }

  @Mock lateinit var animationBackend: AnimationBackend
  @Mock lateinit var bitmapFramePreparer: BitmapFramePreparer
  @Mock lateinit var bitmapFrameCache: BitmapFrameCache
  @Mock lateinit var onAnimationLoaded: () -> Unit

  private lateinit var bitmapFramePreparationStrategy: BitmapFramePreparationStrategy

  @Before
  fun setup() {
    MockitoAnnotations.initMocks(this)
    bitmapFramePreparationStrategy =
        FixedNumberBitmapFramePreparationStrategy(NUMBER_OF_FRAMES_TO_PREPARE)
    whenever(animationBackend.frameCount).thenReturn(FRAME_COUNT)
    whenever(bitmapFramePreparer.prepareFrame(eq(bitmapFrameCache), eq(animationBackend), any()))
        .thenReturn(true)
  }

  @Test
  fun testPrepareFrames_FromFirstFrame() {
    bitmapFramePreparationStrategy.prepareFrames(
        bitmapFramePreparer, bitmapFrameCache, animationBackend, 0, onAnimationLoaded)
    verifyPrepareCalledForFramesInOrder(1, 2, 3)
  }

  @Test
  fun testPrepareFrames_FromLastFrame() {
    bitmapFramePreparationStrategy.prepareFrames(
        bitmapFramePreparer, bitmapFrameCache, animationBackend, 9, onAnimationLoaded)
    verifyPrepareCalledForFramesInOrder(0, 1, 2)
  }

  @Test
  fun testPrepareFrames_ExactlyLastFrames() {
    bitmapFramePreparationStrategy.prepareFrames(
        bitmapFramePreparer, bitmapFrameCache, animationBackend, 6, onAnimationLoaded)
    verifyPrepareCalledForFramesInOrder(7, 8, 9)
  }

  @Test
  fun testPrepareFrames_FrameOverflow() {
    bitmapFramePreparationStrategy.prepareFrames(
        bitmapFramePreparer, bitmapFrameCache, animationBackend, 8, onAnimationLoaded)
    verifyPrepareCalledForFramesInOrder(9, 0, 1)
  }

  @Test
  fun testPrepareFrames_FromFirstFrame_WhenBitmapFramePreparerAlwaysFails() {
    whenever(bitmapFramePreparer.prepareFrame(eq(bitmapFrameCache), eq(animationBackend), any()))
        .thenReturn(false)
    bitmapFramePreparationStrategy.prepareFrames(
        bitmapFramePreparer, bitmapFrameCache, animationBackend, 0, onAnimationLoaded)
    verifyPrepareCalledForFramesInOrder(1)
  }

  @Test
  fun testPrepareFrames_FromFirstFrame_WhenBitmapFramePreparerFailsForSelectedFrames() {
    whenever(bitmapFramePreparer.prepareFrame(eq(bitmapFrameCache), eq(animationBackend), eq(2)))
        .thenReturn(false)
    whenever(bitmapFramePreparer.prepareFrame(eq(bitmapFrameCache), eq(animationBackend), eq(3)))
        .thenReturn(false)
    bitmapFramePreparationStrategy.prepareFrames(
        bitmapFramePreparer, bitmapFrameCache, animationBackend, 0, onAnimationLoaded)
    verifyPrepareCalledForFramesInOrder(1, 2)
  }

  @Test
  fun testPrepareFrames_onAnimationLoadedIsTrigger_WhenFramesAreLoaded() {
    bitmapFramePreparationStrategy.prepareFrames(
        bitmapFramePreparer, bitmapFrameCache, animationBackend, 0, onAnimationLoaded)

    verify(onAnimationLoaded).invoke()
  }

  private fun verifyPrepareCalledForFramesInOrder(vararg frameNumbers: Int) {
    val inOrderBitmapFramePreparer = inOrder(bitmapFramePreparer)
    for (frameNumber in frameNumbers) {
      inOrderBitmapFramePreparer
          .verify(bitmapFramePreparer)
          .prepareFrame(bitmapFrameCache, animationBackend, frameNumber)
    }
    inOrderBitmapFramePreparer.verifyNoMoreInteractions()
  }
}
